src/zlib_stream.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019-2021 libsbn project contributors.
+// libsbn is free software under the GPLv3; see LICENSE file for details.
+
+#include "zlib_stream.hpp"
+
+namespace zlib {
+
+namespace detail {
+
+constexpr Result::Code call_zlib(int zlib_code, const z_stream& impl) {
+  if (zlib_code < 0) {
+    const auto code = static_cast<typename Exception::Code>(zlib_code);
+    if (impl.msg) {
+      throw Exception{code, impl.msg};
+    } else {
+      throw Exception{code};
+    }
+  }
+  return static_cast<Result::Code>(zlib_code);
+}
+
+}  // namespace detail
+
+ZStream::ZStream() {
+  impl_.next_in = nullptr;
+  impl_.avail_in = 0;
+  impl_.zalloc = nullptr;
+  impl_.zfree = nullptr;
+  impl_.opaque = this;
+  auto ret = detail::call_zlib(::inflateInit2(&impl_, 16 + MAX_WBITS), impl_);
+  if (ret != Result::Code::ok) {
+    throw std::logic_error("Unexpected result from zlib");
+  }
+}
+
+ZStream::~ZStream() try { close(); } catch (...) {
+}
+
+void ZStream::close() {
+  if (!closed_.test_and_set()) {
+    auto ret = detail::call_zlib(::inflateEnd(&impl_), impl_);
+    if (ret != Result::Code::ok) {
+      throw std::logic_error("Unexpected result from zlib");
```

Comment:
Does this throw get silenced by the above catch?

Reply:
When Close() is called from the destructor the exception is silenced. But if it gets manually called somewhere else it should still be able to throw and allow the caller to handle the exception.